src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -17903,5 +17922,1075 @@ public class XAttribute : System.Attribute
+            string source = """
+                using System.Collections;
+                using System.Collections.Generic;
+
+                MyCollection x = [uint.MaxValue];
+
+                class MyCollection : IEnumerable<uint>
+                {
+                    private List<uint> _list = new();
+                    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
+                    IEnumerator<uint> IEnumerable<uint>.GetEnumerator() => _list.GetEnumerator();
+                    public void Add(sbyte s) { _list.Add((uint)s); }
+                }
+                """;
+
+            var comp = CreateCompilation(new[] { source, s_collectionExtensions }, targetFramework: TargetFramework.Net70);
+            comp.VerifyEmitDiagnostics(
+                // 0.cs(4,19): error CS1950: The best overloaded Add method 'MyCollection.Add(sbyte)' for the collection initializer has some invalid arguments
+                // MyCollection x = [uint.MaxValue];
+                Diagnostic(ErrorCode.ERR_BadArgTypesForCollectionAdd, "uint.MaxValue").WithArguments("MyCollection.Add(sbyte)").WithLocation(4, 19),
+                // 0.cs(4,19): error CS1503: Argument 1: cannot convert from 'uint' to 'sbyte'
+                // MyCollection x = [uint.MaxValue];
+                Diagnostic(ErrorCode.ERR_BadArgType, "uint.MaxValue").WithArguments("1", "uint", "sbyte").WithLocation(4, 19)
+                );
+        }
+
+        [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/69521")]
+        [InlineData("IEnumerable<int>, IEnumerable<string>")]
+        [InlineData("IEnumerable<string>, IEnumerable<int>")]
+        [InlineData("IString, IInteger")]
+        [InlineData("IInteger, IString")]
+        public void ConvertToCorrectIEnumerable(string interfaces)
+        {
+            string source = $$"""
+                using System.Collections;
+                using System.Collections.Generic;
+
+                Collection c = [1];
+
+                interface IString : IEnumerable<string> { }
+                interface IInteger : IEnumerable<int> { }
+
+                class Collection : {{interfaces}}
+                {
+                    IEnumerator IEnumerable.GetEnumerator() => null;
+                    IEnumerator<string> IEnumerable<string>.GetEnumerator() => null;
+                    IEnumerator<int> IEnumerable<int>.GetEnumerator() => null;
+
+                    public void Add(int i) { System.Console.Write("RAN"); }
+                    public void Add(string s) => throw null;
+                }
+                """;
+
+            var comp = CreateCompilation(source).VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "RAN");
+        }
+
+        [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/69521")]
+        [InlineData("IEnumerable<int>, IEnumerable<long>")]
```

Comment:
Let's have an example with unrelated types, both being correct. `Collection` implements `IEnumerable<I1>` and `IEnumerable<I2>`, perhaps, and `C` implements both `I1` and `I2`. It would also be good to have some tests with user-defined conversions:

1. User defined conversion (explicit and implicit) on the originating type to the `T` in the enumerable
2. User defined conversion (explicit and implicit) on the `T`, from the originating type.

I don't expect the explicit scenarios to work, just good to test.